File: android/app/src/main/java/com/rbdarts/core/observability/Diagnostics.kt
package com.rbdarts.core.observability

enum class DiagnosticSeverity { Debug, Info, Warning, Error }

data class DiagnosticEvent(
    val name: String,
    val severity: DiagnosticSeverity,
    val attributes: Map<String, String> = emptyMap()
)

object DiagnosticEventNames {
    const val ScoreEntryRejected = "score_entry_rejected"
    const val CorrectionRequested = "correction_requested"
    const val CorrectionRejected = "correction_rejected"
    const val CorrectionApplied = "correction_applied"
    const val GameLocked = "game_locked"
    const val GameUnlockRequested = "game_unlock_requested"
    const val SubstitutionApplied = "substitution_applied"
    const val BaseballStatsViewed = "baseball_stats_viewed"
    const val BaseballProjectionUnavailable = "baseball_projection_unavailable"
}

interface DiagnosticsReporter {
    fun record(event: DiagnosticEvent)
}

class PrivacySafeDiagnostics : DiagnosticsReporter {
    private val redactedKeys = listOf("token", "secret", "password", "email", "reason", "providerUserId")
    val events: MutableList<DiagnosticEvent> = mutableListOf()

    override fun record(event: DiagnosticEvent) {
        events += event.copy(
            attributes = event.attributes.filterKeys { key ->
                redactedKeys.none { forbidden -> key.contains(forbidden, ignoreCase = true) }
            }
        )
    }
}

File: ios/RBDarts/Shared/Observability/Diagnostics.swift
import Foundation

public enum DiagnosticSeverity: String, Sendable {
    case debug
    case info
    case warning
    case error
}

public struct DiagnosticEvent: Equatable, Sendable {
    public let name: String
    public let severity: DiagnosticSeverity
    public let attributes: [String: String]
}

public enum DiagnosticEventName {
    public static let scoreEntryRejected = "score_entry_rejected"
    public static let correctionRequested = "correction_requested"
    public static let correctionRejected = "correction_rejected"
    public static let correctionApplied = "correction_applied"
    public static let gameLocked = "game_locked"
    public static let gameUnlockRequested = "game_unlock_requested"
    public static let substitutionApplied = "substitution_applied"
    public static let baseballStatsViewed = "baseball_stats_viewed"
    public static let baseballProjectionUnavailable = "baseball_projection_unavailable"
}

public protocol DiagnosticsReporting: Sendable {
    func record(_ event: DiagnosticEvent)
}

public final class PrivacySafeDiagnostics: DiagnosticsReporting, @unchecked Sendable {
    public private(set) var events: [DiagnosticEvent] = []
    private let redactedKeys = ["token", "secret", "password", "email", "reason", "providerUserId"]

    public init() {}

    public func record(_ event: DiagnosticEvent) {
        let filtered = event.attributes.filter { key, _ in
            !redactedKeys.contains { key.localizedCaseInsensitiveContains($0) }
        }
        events.append(DiagnosticEvent(name: event.name, severity: event.severity, attributes: filtered))
    }
}
